### PR TITLE
Add a boundary check before fuse read

### LINKED ISSUE #13050: alluxio fuse read data error
先通过alluxio fuse mount alluxio的目录， 然后上层通过go read file， alluxio fuse进程显示异常如下
java.lang.IndexOutOfBoundsException: Index: 4, Size: 4
        at java.util.ArrayList.rangeCheck(ArrayList.java:659)
        at java.util.ArrayList.get(ArrayList.java:435)
        at alluxio.client.file.AlluxioFileInStream.triggerAsyncCaching(AlluxioFileInStream.java:389)
        at alluxio.client.file.AlluxioFileInStream.closeBlockInStream(AlluxioFileInStream.java:381)
        at alluxio.client.file.AlluxioFileInStream.read(AlluxioFileInStream.java:198)
        at alluxio.fuse.AlluxioJniFuseFileSystem.readInternal(AlluxioJniFuseFileSystem.java:336)
        at alluxio.fuse.AlluxioJniFuseFileSystem.lambda$read$5(AlluxioJniFuseFileSystem.java:317)
        at alluxio.fuse.AlluxioFuseUtils.call(AlluxioFuseUtils.java:245)
        at alluxio.fuse.AlluxioJniFuseFileSystem.read(AlluxioJniFuseFileSystem.java:317)
        at alluxio.jnifuse.AbstractFuseFileSystem.readCallback(AbstractFuseFileSystem.java:150)
2021-03-12 04:22:39,116 ERROR AlluxioJniFuseFileSystem - Failed to read, path: /data/0.5G_1.txt size: 4096 offset: 536866816

using jnr fuse， also find the error
2021-03-12 08:36:48,521 ERROR AlluxioFuseFileSystem - Failed to read file=/data/0.5G_9.txt, offset=536862720, size=8192
java.lang.IndexOutOfBoundsException: Index: 4, Size: 4
        at java.util.ArrayList.rangeCheck(ArrayList.java:659)
        at java.util.ArrayList.get(ArrayList.java:435)
        at alluxio.client.file.AlluxioFileInStream.triggerAsyncCaching(AlluxioFileInStream.java:389)
        at alluxio.client.file.AlluxioFileInStream.closeBlockInStream(AlluxioFileInStream.java:381)
        at alluxio.client.file.AlluxioFileInStream.read(AlluxioFileInStream.java:198)
        at alluxio.fuse.AlluxioFuseFileSystem.readInternal(AlluxioFuseFileSystem.java:624)
        at alluxio.fuse.AlluxioFuseFileSystem.lambda$read$7(AlluxioFuseFileSystem.java:596)
        at alluxio.fuse.AlluxioFuseUtils.call(AlluxioFuseUtils.java:245)
        at alluxio.fuse.AlluxioFuseFileSystem.read(AlluxioFuseFileSystem.java:596)
        at ru.serce.jnrfuse.AbstractFuseFS.lambda$init$3(AbstractFuseFS.java:135)
        at jnr.ffi.provider.jffi.NativeClosureProxy$$impl$$9.invoke(Unknown Source)
unique: 2465256, opcode: FLUSH (25), nodeid: 594, insize: 64, pid: 777281

到文件末尾必现：
tail -n 100  /data/benchmark/data/0.5G_9.txt 
tail: error reading '/data/benchmark/data/0.5G_9.txt': Bad message

### PR DESCRIPTION
**Pull request by maobaolong**
Fix #13050

### CODE REVIEW

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 1]
```diff
@@ -327,11 +327,13 @@ private int readInternal(String path, ByteBuffer buf, long size, long offset, Fu
           LOG.error("Cannot find fd {} for {}", fd, path);
           return -ErrorCodes.EBADFD();
         }
-        is.seek(offset);
-        while (rd >= 0 && nread < size) {
-          rd = is.read(dest, nread, sz - nread);
-          if (rd >= 0) {
-            nread += rd;
+        if (offset - is.getPos() < is.remaining()) {
```
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:330]: if offset exceeds the input stream length, according to POSIX definition, we shall still read bytes until file ends.

**Review by apc999 — CHANGES_REQUESTED** [review]
Can you add a unit test at AlluxioJniFuseFileSystemTest too?

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java (hunk 1 above)
Comment (maobaolong) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:330]: I think if offset is the start point which the current request want to read, but if it exceeds the input stream length, it means no bytes to read, correct me if I'm wrong?

**Comment by apc999** [comment]
@maobaolong 
can you add a UT so that when passing offset more than file length, read 0 bytes?

**Comment by apc999** [comment]
@maobaolong  possibly to add your test to AlluxioJniFuseFileSystemTest too?

File: integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java [hunk 2]
```diff
@@ -415,6 +417,88 @@ public void read() throws Exception {
     assertArrayEquals("Source and dst data should be equal", expected, dst);
   }
 
+  public void readOffset() throws Exception {
```
Comment (apc999) [inline integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java:420]: `@Test`? or the tests will not run

**Review by apc999 — APPROVED** [review]
one comment left . otherwise LGTM

File: integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java (hunk 2 above)
Comment (maobaolong) [reply integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java:421]: Sure, sorry for miss it.

**Pull request merged**